### PR TITLE
Use base64 encoded length for decoder table size calculation of binary headers

### PR DESCRIPTION
**Pull request by yang-g**

### CODE REVIEW

File: src/core/lib/transport/metadata.c [hunk 1]
```diff
@@ -93,6 +94,9 @@ typedef struct internal_string {
 
   gpr_slice base64_and_huffman;
 
+  uint8_t has_size_in_decoder_table;
```
Comment (ctiller) [inline src/core/lib/transport/metadata.c]: s/uint8_t/bool - we have C99 now!

Comment (yang-g) [inline src/core/lib/transport/metadata.c]: Done.

**Comment by yang-g** [comment]
Green.

File: src/core/lib/transport/metadata.c [hunk 2]
```diff
@@ -576,6 +581,39 @@ grpc_mdelem *grpc_mdelem_from_string_and_buffer(const char *key,
...
+    if (grpc_is_binary_header(
+            (const char *)GPR_SLICE_START_PTR(elem->key->slice),
+            GPR_SLICE_LENGTH(elem->key->slice))) {
+      return overhead_and_key + get_base64_encoded_size(value_len);
+    } else {
+      return overhead_and_key + value_len;
+    }
+  } else {
+    internal_string *is = (internal_string *)elem->value;
+    gpr_atm current_size = gpr_atm_no_barrier_load(&is->size_in_decoder_table);
```
Comment (ctiller) [inline src/core/lib/transport/metadata.c]: gpr_atm_acq_load

File: src/core/lib/transport/metadata.c [hunk 3]
```diff
@@ -576,6 +581,39 @@ grpc_mdelem *grpc_mdelem_from_string_and_buffer(const char *key,
...
+    if (current_size == SIZE_IN_DECODER_TABLE_NOT_SET) {
+      if (grpc_is_binary_header(
+              (const char *)GPR_SLICE_START_PTR(elem->key->slice),
+              GPR_SLICE_LENGTH(elem->key->slice))) {
+        current_size = (gpr_atm)get_base64_encoded_size(value_len);
+      } else {
+        current_size = (gpr_atm)value_len;
+      }
+      gpr_atm_no_barrier_store(&is->size_in_decoder_table, current_size);
```
Comment (ctiller) [inline src/core/lib/transport/metadata.c]: gpr_atm_rel_store

**Comment by yang-g** [comment]
Done and green enough?

**Comment by yang-g** [comment]
Green again

**Comment by ctiller** [comment]
Lgtm

**Pull request merged**